GH-37484: [Python] Add a FixedSizeTensorScalar class

rok opened the pull request:
### Rationale for this change

When working with `FixedSizeTensorArray` we want to access individual tensors. This would be enabled by adding:
```python
def FixedSizeTensorScalar(pa.ExtensionScalar):
    def to_numpy_ndarray(): ...
```

See #37484.

### What changes are included in this PR?

This adds `FixedSizeTensorScalar` and tests for it.

### Are there any user-facing changes?

Yes, when calling `FixedSizeTensorArray[i]` we would get back `FixedSizeTensorScalar` instead of `ExtensionScalar`.
* Closes: #37484

alippai commented:
Would the numpy array api or https://data-apis.org/array-api/latest/purpose_and_scope.html add any value here?

rok commented:
@alippai This PR would effectively implement a [`__pos__(i)`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__pos__.html) method. Which should per `array-api` return an array while pyarrow typically returns scalar. I'm not sure how to reconcile this. We should probably have a broader discussion on the mailing list about adopting array API.

AlenkaF commented on python/pyarrow/types.pxi:
```diff
@@ -1629,6 +1629,9 @@ cdef class FixedShapeTensorType(BaseExtensionType):
                                     self.dim_names, self.permutation)
 
+    def __arrow_ext_scalar_class__(self):
+        return FixedShapeTensorScalarType
```
```suggestion
        return FixedShapeTensorScalar
```

AlenkaF reviewed (commented):
Thanks for working on this! Added two suggestions, otherwise the Pyhton part LGTM.

AlenkaF commented:
> Would the numpy array api or https://data-apis.org/array-api/latest/purpose_and_scope.html add any value here?

I am very much hoping we could implement DLPack in Arrow: https://github.com/apache/arrow/issues/33984. Specially for the new tensor arrays, it would be very beneficial!

> @alippai This PR would effectively implement a [`__pos__(i)`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__pos__.html) method. Which should per `array-api` return an array while pyarrow typically returns scalar. I'm not sure how to reconcile this. We should probably have a broader discussion on the mailing list about adopting array API.

+1

AlenkaF commented:
One more thing, can the change in the C++ code (`GetTensor` feature) also be reflected in the title of the PR and the description? I am not sure how we want to rename it, so I haven't made any change.

rok commented on python/pyarrow/array.pxi:
```diff
@@ -3325,6 +3325,26 @@ cdef class ExtensionArray(Array):
         return result
 
 
+class FixedShapeTensorScalarType(ExtensionScalar):
+    """
+    Concrete class for fixed shape tensor extension scalar type.
+    """
```
Oh, indeed, I'll move it!

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -3519,16 +3519,32 @@ class FixedShapeTensorArray(ExtensionArray):
+            CResult[shared_ptr[CTensor]] ctensor
+        with nogil:
+            ctensor = ext_array.ToTensor()
+        return pyarrow_wrap_tensor(GetResultValue(ctensor)).to_numpy()
```
I don't remember exactly what ToTensor does by default with regards to permutation (assuming it takes it into account), but so this expands the behaviour of `to_numpy_ndarray`, I think. In that case, can you expand the docstring and add some tests for this?

rok replied:
`ToTensor` uses permutation to calculate strides and zero-copy creates `pyarrow.Tensor`.
`pyarrow.Tensor` can be zero-copy converted to a numpy array. I am not sure what that conversion does with strides, but we should check.
Updated the docstring.

jorisvandenbossche reviewed (commented):
Didn't yet look in detail, but added some quick drive-by comments. And thanks for working on this!

Can you also add some tests for the new Scalar class?

Currently, for the Python bindings, you added a `get_tensor(i)` method on the array class, but wouldn't make sense to (also/instead) add a `to_tensor()` method on the scalar class, since this is to get a Tensor for a single element (scalar) of the array?

rok commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -82,6 +82,45 @@ Status ComputeStrides(const FixedWidthType& type, const std::vector<int64_t>& sh
+  std::vector<int64_t> strides;
+  ARROW_CHECK_OK(ComputeStrides(*value_type.get(), shape, permutation, &strides));
+
+  // TODO: can this be done without copying?
```
I believe not, removing comment.

rok commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -82,6 +82,45 @@ Status ComputeStrides(const FixedWidthType& type, const std::vector<int64_t>& sh
+  ARROW_CHECK_OK(ComputeStrides(*value_type.get(), shape, permutation, &strides));
+
+  // TODO: can this be done without copying?
+  auto list_arr = ext_arr->value_slice(i)->data();
```
Indeed. Changed.

rok commented:
> Didn't yet look in detail, but added some quick drive-by comments. And thanks for working on this!

Thanks for the review @jorisvandenbossche ! I've addressed your points.

> Can you also add some tests for the new Scalar class?

Will do. I'll try to reuse the ones I'm adding in https://github.com/apache/arrow/pull/38008/files

> Currently, for the Python bindings, you added a `get_tensor(i)` method on the array class, but wouldn't make sense to (also/instead) add a `to_tensor()` method on the scalar class, since this is to get a Tensor for a single element (scalar) of the array?

Indeed, added. I've also made the `get_tensor` logic in c++ be type method that takes a scalar as opposed to array method that takes an index so that it's easier to call from Python. (This cam  out of a discussion in https://github.com/apache/arrow/pull/38008)

jorisvandenbossche commented on python/pyarrow/includes/libarrow.pxd:
```diff
@@ -2650,24 +2650,25 @@ cdef extern from "arrow/extension_type.h" namespace "arrow":
 cdef extern from "arrow/extension/fixed_shape_tensor.h" namespace "arrow::extension":
     cdef cppclass CFixedShapeTensorType \
-            " arrow::extension::FixedShapeTensorType"(CExtensionType):
+            " arrow::extension::FixedShapeTensorType"(CExtensionType) nogil:
```
I think you can add the nogil to the `cdef extern .. namespace ..` line just above (then you don't need to add it to both classes individually)

rok replied:
Nice! Changed.

jorisvandenbossche commented on python/pyarrow/scalar.pxi:
```diff
@@ -1027,6 +1027,32 @@ cdef class ExtensionScalar(Scalar):
+
+    def to_tensor(self):
+        """
+        Convert fixed shape tensor extension scalar to a pyarrow.Tensor.
```
Can you add a Returns section? (then that will link to the pyarrow.Tensor docs I think)

rok replied:
Done.

pitrou replied:
Not done.

rok replied:
Thanks for catching that! I must have gotten distracted by `FixedShapeTensorArray.to_tensor()`. Hopefully done now.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1319,38 +1319,65 @@ def test_tensor_type():
+    from numpy.lib.stride_tricks import as_strided
+
+    values = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
+    np_arr = np.array(values, dtype=np.int8)
```
The tensor is using float32?

rok replied:
Indeed. I parameterized the test so these are no longer static.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1319,38 +1319,65 @@ def test_tensor_type():
     tensor_type = pa.fixed_shape_tensor(pa.float32(), [2, 3])
-    storage = pa.array([[1, 2, 3, 4, 5, 6], [1, 2, 3, 4, 5, 6]],
-                       pa.list_(pa.float32(), 6))
+    storage = pa.array(np_arr.reshape(2, 6).tolist(), pa.list_(pa.float32(), 6))
```
I personally find the previous way easier to read. I think it's fine to repeat the values for readability.

rok replied:
Reverted.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1319,38 +1319,65 @@ def test_tensor_type():
+    # expected = np.array([[[7, 8, 9], [10, 11, 12]]], dtype=np.float32)
     result = arr[:1].to_numpy_ndarray()
-    np.testing.assert_array_equal(result, expected)
+    # np.testing.assert_array_equal(result, expected)
```
uncomment?

rok replied:
Offset of sliced array is not correctly handled yet. Still needs to be solved.

rok replied:
@pitrou would a buffer of an array sliced in python have correct `offset` accessible in C++? I've tried slicing it with `ext_arr->Slice(ext_arr->offset(), ..)` but only got data with 0 offset.
https://github.com/apache/arrow/blob/5085b097ce672c624c9b88f77f2def6d371a7464/cpp/src/arrow/extension/fixed_shape_tensor.cc#L372-L375
This approach worked with a Scalar:
https://github.com/apache/arrow/blob/5085b097ce672c624c9b88f77f2def6d371a7464/cpp/src/arrow/extension/fixed_shape_tensor.cc#L239-L243

jorisvandenbossche replied:
Didn't check in detail if that's the issue here, but pay attention to the offset of the ListArray versus the offset of the underlying values array (which then further depends on whether you ask for `values()` or `Flatten()`). 

Python example (which I think should translate to C++ more or less directly):

```
In [31]: arr = pa.array([[1, 2], [3, 4]], pa.list_(pa.int64(), 2))

In [32]: arr.slice(1, 1)
Out[32]: 
<pyarrow.lib.FixedSizeListArray object at 0x7f283ce57d00>
[
  [
    3,
    4
  ]
]

In [33]: arr.slice(1, 1).offset
Out[33]: 1

In [34]: arr.slice(1, 1).flatten()
Out[34]: 
<pyarrow.lib.Int64Array object at 0x7f2836df4fa0>
[
  3,
  4
]

In [35]: arr.slice(1, 1).flatten().offset
Out[35]: 2

In [36]: arr.slice(1, 1).values
Out[36]: 
<pyarrow.lib.Int64Array object at 0x7f283ce56200>
[
  1,
  2,
  3,
  4
]

In [37]: arr.slice(1, 1).values.offset
Out[37]: 0
```

rok replied:
Thanks for the suggestion! I did have to slice manually but it did the trick.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1319,38 +1319,65 @@ def test_tensor_type():
         pa.FixedShapeTensorArray.from_numpy_ndarray(arr)
 
+    bw = np.int8().itemsize
+    arr = as_strided(np_arr, shape=(3, 4), strides=(bw * 4, bw), writeable=False)
```
```suggestion
    arr = np_arr.reshape(3, 4)
```

?

rok replied:
That's more readable yeah, changing.

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1319,38 +1319,65 @@ def test_tensor_type():
+    arr = as_strided(np_arr, shape=(3, 4), strides=(bw * 4, bw), writeable=False)
+    tensor_array_from_numpy = pa.FixedShapeTensorArray.from_numpy_ndarray(arr)
+    assert tensor_array_from_numpy.type.shape == [4]
+    assert tensor_array_from_numpy.type.permutation == [1]
```
The spec says: "The indices contain a permutation of the values [0, 1, .., N-1] where N is the number of dimensions.", so how can this be [1] ?

rok replied:
Fixed, this would now return [0].

jorisvandenbossche replied:
On main, it seems that the type constructor also happily lets you create a type with an invalid permutation:

```
In [5]: pa.fixed_shape_tensor(pa.int8(), shape=(4, ), permutation=[3])
Out[5]: FixedShapeTensorType(extension<arrow.fixed_shape_tensor[value_type=int8, shape=[4], permutation=[3]]>)
```

We should probably add some validation to this type factory function to ensure the permutation is valid?

rok replied:
Added a check to `FixedShapeTensorType::Make`:
```cpp
  for (auto i : permutation) {
    if (i < 0 || i >= static_cast<int64_t>(shape.size())) {
      return Status::Invalid("permutation indices must be in [0, shape.size()). Got: ",
                             i);
    }
  }
```

jorisvandenbossche replied:
Thanks!

jorisvandenbossche commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1319,38 +1319,65 @@ def test_tensor_type():
-        [[[1, 2, 3], [4, 5, 6]], [[1, 2, 3], [4, 5, 6]]],
+        [[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]],
         dtype=np.float32, order="C")
     tensor_array_from_numpy = pa.FixedShapeTensorArray.from_numpy_ndarray(arr)
```
I know this was already mixed up now, but since you are expanding the tests, can you split it? For example separate the `from_numpy_ndarray` parts into their own test?

rok replied:
Done.

jorisvandenbossche commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -203,6 +215,40 @@ std::shared_ptr<Array> FixedShapeTensorType::MakeArray(
+  const auto array =
+      internal::checked_pointer_cast<const FixedSizeListScalar>(scalar->value)->value;
+  const auto value_type =
+      internal::checked_pointer_cast<FixedWidthType>(this->value_type());
```
This should first validate that it is indeed a fixed width type (and if not raise a proper error message) instead of just casting? 
Also, it should probably check that there are no nulls?

(and add tests for this)

rok replied:
Added checks and tests.

jorisvandenbossche commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -298,25 +344,25 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
-  std::vector<int64_t> shape;
-  for (int64_t& i : permutation) {
-    shape.emplace_back(ext_type->shape()[i]);
-    ++i;
```
(the C++ tests are also failing, potentially related)

rok replied:
Reverted.

jorisvandenbossche reviewed (commented):
A more thorough review round this time. There are still a bunch of TODOs about strides not yet correctly handled. Are those out of date or does that still need to be implemented?

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -3518,17 +3518,38 @@ class FixedShapeTensorArray(ExtensionArray):
+
+        Parameters
+        ----------
+        i : int64_t
```
```suggestion
        i : int
```

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -3518,17 +3518,38 @@ class FixedShapeTensorArray(ExtensionArray):
+        ----------
+        i : int64_t
+            The index of the tensor to get.
+        Returns
```
```suggestion

        Returns
```

rok commented on python/pyarrow/array.pxi:
```diff
@@ -3518,17 +3518,38 @@ class FixedShapeTensorArray(ExtensionArray):
+        """
+        return self.to_tensor().to_numpy()
+
+    def to_tensor(self):
```
Added.

rok commented on python/pyarrow/array.pxi:
```diff
@@ -3518,17 +3518,38 @@ class FixedShapeTensorArray(ExtensionArray):
-        Convert fixed shape tensor extension array to a numpy array (with dim+1).
+        Convert fixed shape tensor extension array to a numpy.ndarray with zero copy.
+        First dimension of ndarray will be the length of the fixed shape tensor array
+        and the rest of the dimensions will match the shape of the fixed shape tensor.
```
Updated to state it will be permuted (except for the first dimension).

rok commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -321,6 +324,70 @@ TEST_F(TestExtensionType, TestFromTensorType) {
+  auto tensor_dim_names = std::vector<std::vector<std::string>>{
+      {"", "y", "z"},      {"", "y", "z"},     {"", "y", "z"}, {"", "y", "z"},
+      {"", "y", "z"},      {"", "y", "z"},     {"", "y", "z"}, {"", "y", "z"},
+      {"", "H", "W", "C"}, {"", "H", "W", "C"}};
```
I've updated the tests as per my understanding. Please do check. Doc is not super clear indeed.

jorisvandenbossche replied:
> I've updated the tests as per my understanding. Please do check. Doc is not super clear indeed.

But a ToTensor implementation already existed before as well (for the full array), no? What did it do there?

rok replied:
This PR now doesn't change the previous behavior. `ToTensor` uses the permutation to construct a Tensor of permuted shape and strides to match the shape and data buffer. This now introduces a minor change in the way dim_names are composed and how permutations are done and adds tests. Previous implementation had a bug if `ext_type->permutation();` was empty etc.

jorisvandenbossche replied:
"This PR now doesn't change the previous behavior" and "This now introduces a minor change in the way dim_names are composed" seems to contradict each other?

rok replied:
Sorry I was referring only to the way strides and shapes were produced.

jorisvandenbossche replied:
So the behaviour of ToTensor changed regarding the returned shape? (i.e. the implementation in main is wrong?)

Can you show a before/after comparison with an example?

rok replied:
> So the behaviour of ToTensor changed regarding the returned shape? (i.e. the implementation in main is wrong?)
> Can you show a before/after comparison with an example?

The implementation was wrong for cases where `ext_type->permutation()` was empty but `ext_type->dim_names()` was not as it would not correctly assemble `dim_names` for the new `Tensor`. Other changes are more implementation details, e.g. using `internal::Permute<std::string>(permutation, &dim_names);` instead of a for loop. The way strides and shapes are permuted at `ToTensor` is (currently) not changed to main.

rok replied:
Are you saying you don't think the behavior as shown by the test is correct?

rok commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -344,6 +411,9 @@ TEST_F(TestExtensionType, RoundtripTensor) {
       {48, 8, 24}, {144, 48, 8}, {144, 8, 48}, {144, 48, 24, 8}, {144, 8, 24, 48}};
+  auto permutations =
+      std::vector<std::vector<int64_t>>{{0, 1}, {1, 0}, {0, 1}, {1, 0},    {0, 1},
+                                        {1, 0}, {0, 1}, {0, 1}, {0, 1, 2}, {2, 1, 0}};
```
Removed.

rok commented on cpp/src/arrow/extension/fixed_shape_tensor.h:
```diff
@@ -43,6 +43,14 @@ class ARROW_EXPORT FixedShapeTensorArray : public ExtensionArray {
+  /// derived from shape and permutation of FixedShapeTensorArray. Shape and
+  /// dim_names will be permuted according to permutation stored in the
+  /// FixedShapeTensorArray metadata.
+  const Result<std::shared_ptr<Tensor>> GetTensor(const int64_t i) const;
```
Perhaps it's redundant indeed. Removed here (FSTArray) and in Python API (from FSTArray). Now only `GetTensor(scalar)` is available (for both c++ and Python).

rok commented on cpp/src/arrow/extension/fixed_shape_tensor.h:
```diff
@@ -94,6 +102,10 @@ class ARROW_EXPORT FixedShapeTensorType : public ExtensionType {
   std::shared_ptr<Array> MakeArray(std::shared_ptr<ArrayData> data) const override;
 
+  /// \brief Create a FixedShapeTensorArray from a Scalar Tensor
+  const Result<std::shared_ptr<Tensor>> GetTensor(
```
Updated.

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -3573,17 +3595,20 @@ class FixedShapeTensorArray(ExtensionArray):
-        shape = obj.shape[1:]
-        size = obj.size / obj.shape[0]
+        shape = np.take(obj.shape, permutation)
+        values = np.lib.stride_tricks.as_strided(
+            obj, shape=(obj.size,), strides=(obj.dtype.itemsize,))
```
Is this different from the ravel that was used before? (or maybe you need `ravel(.., order="K")` to just flatten following the memory layout

rok replied:
The reason for the change was that ravel states in the doc: `[..] A copy is made only if needed. [..]` while `as_strided` only creates a view. I would prefer `ravel` as it's less error prone but I can't immediately see from [source](https://github.com/numpy/numpy/blob/v1.26.0/numpy/core/fromnumeric.py#L1768-L1874) when it would incur a copy.

jorisvandenbossche replied:
In general it will make a copy when the memory layout changes, but by specifying `order="K"`, you say to use the original memory layout, which I think means it will always be zero-copy

rok replied:
I've not seen zero-copy specifically promised for `order="K"`, but it seem it's meant for that. Changed to `ravel`.

rok commented:
@jorisvandenbossche I think slicing should now be resolved. Could you please do another pass?

jorisvandenbossche commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,40 +335,50 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
-                  Status::Invalid(ext_arr->value_type()->ToString(),
-                                  " is not valid data type for a tensor"));
-  auto permutation = ext_type->permutation();
+  const auto storage_array =
+      internal::checked_pointer_cast<FixedSizeListArray>(this->storage());
+  ARROW_ASSIGN_OR_RAISE(const auto flattened_storage_array, storage_array->Flatten());
```
It might be better to move this back after the validation of the type (and add a validation that there are no nulls), because `Flatten()` can incur a copy of the data, and then you would unnecessarily already do a copy when the next type check would fail.

rok replied:
Good point. Moved.

rok commented:
@pitrou :)

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -203,6 +205,46 @@ std::shared_ptr<Array> FixedShapeTensorType::MakeArray(
+const Result<std::shared_ptr<Tensor>> FixedShapeTensorType::GetTensor(
+    const std::shared_ptr<ExtensionScalar>& scalar) const {
+  if (!is_fixed_width(*this->value_type())) {
+    return Status::Invalid("Cannot convert non-fixed-width values to Tensor.");
```
```suggestion
    return Status::TypeError("Cannot convert non-fixed-width values to Tensor.");
```

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -203,6 +205,46 @@ std::shared_ptr<Array> FixedShapeTensorType::MakeArray(
+  const auto array =
+      internal::checked_pointer_cast<const FixedSizeListScalar>(scalar->value)->value;
+  if (array->null_count() > 0) {
+    return Status::Invalid("Cannot convert data with nulls values to Tensor.");
```
```suggestion
    return Status::Invalid("Cannot convert data with nulls to Tensor.");
```

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -203,6 +205,46 @@ std::shared_ptr<Array> FixedShapeTensorType::MakeArray(
+
+  std::vector<int64_t> permutation = this->permutation();
+  if (permutation.empty()) {
+    for (int64_t j = 0; j < static_cast<int64_t>(this->ndim()); ++j) {
+      permutation.emplace_back(j);
+    }
```
```suggestion
    permutation.resize(ndim);
    std::iota(permutation.begin(), permutation.end(), 0);
```

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,40 +335,49 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
+  const auto value_type = ext_type->value_type();
+  ARROW_RETURN_IF(
+      !is_fixed_width(*value_type),
+      Status::Invalid(value_type->ToString(), " is not valid data type for a tensor"));
```
```suggestion
      Status::TypeError(value_type->ToString(), " is not valid data type for a tensor"));
```

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,40 +335,49 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
-    dim_names = {};
   }
+  for (int64_t i = 0; i < static_cast<int64_t>(ext_type->ndim()); i++) {
+    permutation[i] += 1;
```
What? Why?

rok replied:
`ext_type->permutation()` gives us permutation for a single row (range [0-ndim)). We want to create a tensor (ndim+1) from the whole array and we assume the first dimension will always have the greatest stride, so it will get permutation index 0 and remaining values from `ext_type->permutation()` need to be shifted to range [1-ndim+1).

pitrou replied:
I see, can you add a quick explanation as a comment?

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,40 +335,49 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
-                                tensor_strides, dim_names));
-  return tensor;
+      const auto flattened_storage_array,
+      internal::checked_pointer_cast<FixedSizeListArray>(this->storage())->Flatten());
```
Why do you need to call `Flatten`? You can just use the list array's child array...

rok replied:
Done.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -340,6 +391,12 @@ Result<std::shared_ptr<DataType>> FixedShapeTensorType::Make(
+    if (i < 0 || i >= static_cast<int64_t>(shape.size())) {
+      return Status::Invalid("permutation indices must be in [0, shape.size()). Got: ",
+                             i);
+    }
```
Is that all? Am I allowed to pass `[1,2,2,1]` as permutation?

rok replied:
Changed the check and added a test: https://github.com/apache/arrow/pull/37533/commits/37b398e5d266c0543b366513e4a7b42641ca6f94

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -321,6 +322,70 @@ TEST_F(TestExtensionType, TestFromTensorType) {
+                                        {1, 0}, {0, 1}, {1, 0}, {0, 1, 2}, {2, 1, 0}};
+  auto tensor_strides = std::vector<std::vector<int64_t>>{
+      {96, 32, 8}, {96, 8, 24},  {96, 24, 8},  {96, 8, 32},      {48, 24, 8},
+      {48, 8, 16}, {144, 48, 8}, {144, 8, 24}, {144, 48, 24, 8}, {144, 8, 24, 48}};
```
Is it useful to check so many variations? There's not a lot of structure here, and it's not obvious you're trying to exercise something specific.

rok replied:
Done.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -321,6 +322,70 @@ TEST_F(TestExtensionType, TestFromTensorType) {
   }
 }
 
+void CheckToTensor(const std::vector<int64_t> values, const int32_t cell_size,
```
Please pass vectors as const-ref.

rok replied:
Changed.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -321,6 +322,70 @@ TEST_F(TestExtensionType, TestFromTensorType) {
   }
 }
 
+void CheckToTensor(const std::vector<int64_t> values, const int32_t cell_size,
```
Also can we use something else than `int64_t` as data type? So that not all variables here have the same type...

rok replied:
Added a `float32` tests.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -321,6 +322,70 @@ TEST_F(TestExtensionType, TestFromTensorType) {
+
+  auto ext_arr = ExtensionType::WrapArray(ext_type, fsla_arr);
+  const auto tensor_array = std::static_pointer_cast<FixedShapeTensorArray>(ext_arr);
+  ASSERT_OK_AND_ASSIGN(const auto actual_tensor, tensor_array->ToTensor());
```
Why is it called `ToTensor` here but `MakeTensor` for scalars?

rok replied:
Changed.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -321,6 +322,70 @@ TEST_F(TestExtensionType, TestFromTensorType) {
+      {"", "y", "z"},      {"", "y", "z"},     {"", "y", "z"}, {"", "y", "z"},
+      {"", "H", "W", "C"}, {"", "C", "W", "H"}};
+
+  for (size_t i = 0; i < cell_shapes.size(); i++) {
```
Hmm... what is a "cell"? Can we use a terminology that exists in Arrow?

rok replied:
Would scalar be better?

pitrou replied:
"item" or "element" would probably work.

rok replied:
Changed to element.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3519,17 +3519,42 @@ class FixedShapeTensorArray(ExtensionArray):
     def to_numpy_ndarray(self):
         """
-        Convert fixed shape tensor extension array to a numpy array (with dim+1).
+        Convert fixed shape tensor extension array to a numpy.ndarray with zero copy.
```
```suggestion
        Convert fixed shape tensor extension array to a multi-dimensional numpy.ndarray.
        
```

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3519,17 +3519,42 @@ class FixedShapeTensorArray(ExtensionArray):
         """
-        Convert fixed shape tensor extension array to a numpy array (with dim+1).
+        Convert fixed shape tensor extension array to a numpy.ndarray with zero copy.
+        First dimension of ndarray will be the length of the fixed shape tensor array
+        and the rest of the dimensions will match the permuted shape of the fixed
+        shape tensor.
```
```suggestion
        The resulting ndarray will have (ndim + 1) dimensions.
        The size of the first dimension will be the length of the fixed shape tensor array
        and the rest of the dimensions will match the permuted shape of the fixed
        shape tensor.
        
        The conversion is zero-copy.
```

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3519,17 +3519,42 @@ class FixedShapeTensorArray(ExtensionArray):
+
+    def to_tensor(self):
+        """
+        Convert fixed shape tensor extension array to a pyarrow.Tensor with zero copy.
+        First dimension of pyarrow.Tensor will be the length of the fixed shape tensor
+        array and the rest of the dimensions will match the permuted shape of the fixed
+        shape tensor.
+
```
```suggestion
        Convert fixed shape tensor extension array to a pyarrow.Tensor.

        The resulting Tensor will have (ndim + 1) dimensions.
        The size of the first dimension will be the length of the fixed shape tensor array
        and the rest of the dimensions will match the permuted shape of the fixed
        shape tensor.
        
        The conversion is zero-copy.
```

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3573,17 +3595,19 @@ class FixedShapeTensorArray(ExtensionArray):
-                             'C-style contiguous segment.')
+
+        permutation = (-np.array(obj.strides)).argsort()
+        if not permutation[0] == 0:
```
Nit
```suggestion
        if permutation[0] != 0:
```

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3519,17 +3519,42 @@ class FixedShapeTensorArray(ExtensionArray):
 
     def to_numpy_ndarray(self):
```
Why isn't it called `to_numpy` and `from_numpy` btw?

rok replied:
`to_numpy`
```python
import pyarrow as pa
arrow_type = pa.int8()
tensor_type = pa.fixed_shape_tensor(arrow_type, [2, 3])
storage = pa.array([[1, 2, 3, 4, 5, 6], [7, 8, 9, 10, 11, 12]], pa.list_(arrow_type, 6))
arr = pa.ExtensionArray.from_storage(tensor_type, storage)
arr.to_numpy(zero_copy_only=False)
```
already returns storage arrays:
```
array([array([1, 2, 3, 4, 5, 6], dtype=int8),
       array([ 7,  8,  9, 10, 11, 12], dtype=int8)], dtype=object)
```
We agreed (with @jorisvandenbossche  and @AlenkaF ), in a thread I can't find now, to use `to_numpy_ndarray` for getting tensors from `FixedShapeTensorArray` into numpy. I'd use the same convention for `from_numpy_ndarray` to go from `ndarray` into `FixedShapeTensorArray`

pitrou replied:
> ```
> array([array([1, 2, 3, 4, 5, 6], dtype=int8),
>        array([ 7,  8,  9, 10, 11, 12], dtype=int8)], dtype=object)
> ```

In which circumstances is this a desirable behavior? It even makes a copy, right?

rok replied:
I wouldn't know, this is defined in [Array class](https://github.com/apache/arrow/blob/main/python/pyarrow/array.pxi#L1537) and I think we just didn't want to override it.

pitrou replied:
I really think we should change `to_numpy` in this case. The current behavior does not seem useful.

pitrou replied:
Ok, we can make this a separate issue since the function already exists.

rok replied:
Opened https://github.com/apache/arrow/issues/39991.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3573,17 +3595,19 @@ class FixedShapeTensorArray(ExtensionArray):
-            raise ValueError('The data in the numpy array need to be in a single, '
-                             'C-style contiguous segment.')
+
+        permutation = (-np.array(obj.strides)).argsort()
```
Neat, though I think we want a stable sort in case there are duplicates in `strides`:
```suggestion
        permutation = (-np.array(obj.strides)).argsort(kind='stable')
```

rok replied:
Oh, good point!

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -1027,6 +1027,32 @@ cdef class ExtensionScalar(Scalar):
+    Concrete class for fixed shape tensor extension scalar.
+    """
+
+    def to_numpy_ndarray(self):
```
Why not `to_numpy`?

rok replied:
See above.

pitrou replied:
Well, the `Scalar` base class doesn't have a `to_numpy` method...

pitrou commented on python/pyarrow/scalar.pxi:
```diff
@@ -1027,6 +1027,32 @@ cdef class ExtensionScalar(Scalar):
+
+    def to_numpy_ndarray(self):
+        """
+        Convert fixed shape tensor extension scalar to a numpy.ndarray with zero copy.
```
```suggestion
        Convert fixed shape tensor scalar to a numpy.ndarray.
        
        The resulting ndarray's shape matches the permuted shape of the
        fixed shape tensor scalar.
        The conversion is zero-copy.

        Returns
        -------
        numpy.ndarray
```

rok commented on cpp/src/arrow/extension/fixed_shape_tensor.h:
```diff
@@ -94,6 +94,15 @@ class ARROW_EXPORT FixedShapeTensorType : public ExtensionType {
+  /// dim_names will be permuted according to permutation stored in the
+  /// FixedShapeTensorType metadata.
+  const Result<std::shared_ptr<Tensor>> GetTensor(
+      const std::shared_ptr<ExtensionScalar>& scalar) const;
```
Can we access `FixedShapeTensorType->shape()` and `FixedShapeTensorType->permutation()` from a static method?

pitrou replied:
You can get the DataType from the Scalar instance. It will also ensure that the user didn't pass the wrong type by mistake.

rok replied:
Oh, nice. https://github.com/apache/arrow/pull/37533/commits/dd01fc0ef24dcfcd16a692f01468a27f8fa920f7

rok commented:
@pitrou thanks for the review! I think I addressed all your points. Could you please do another pass?

rok commented:
@pitrou could you please do another pass? :)

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,53 +337,83 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
-      dim_names.emplace_back(ext_type->dim_names()[i]);
+  std::vector<int64_t> permutation = ext_type->permutation();
+  if (permutation.empty()) {
+    for (int64_t i = 0; i < static_cast<int64_t>(ext_type->ndim()); i++) {
```
This is `std::iota` like above, right?

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,53 +337,83 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
-  } else {
-    dim_names = {};
   }
+  for (int64_t i = 0; i < static_cast<int64_t>(ext_type->ndim()); i++) {
```
This is going to be executed also if permutation is empty, can you instead give it the right values directly?

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -308,19 +319,89 @@ TEST_F(TestExtensionType, TestFromTensorType) {
+      {"", "y", "z"}, {"", "H", "W", "C"}, {"", "C", "W", "H"}};
+
+  for (size_t i = 0; i < element_shapes.size(); i++) {
+    CheckToTensor<float_t>(values, float32(), element_sizes[i], element_shapes[i],
+                           element_permutations[i], element_dim_names[i],
+                           tensor_shapes[i], tensor_dim_names[i], tensor_strides_32[i]);
+    CheckToTensor(values_, int64(), element_sizes[i], element_shapes[i],
```
Readability: can we avoid having both `values` and `values_`? Also, why are you explicitly specifying `float_t` above?

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -462,4 +543,121 @@ TEST_F(TestExtensionType, ToString) {
   ASSERT_EQ(expected_3, result_3);
 }
 
+TEST_F(TestExtensionType, GetScalar) {
```
Is this test exercising anything specific to fixed_shape_tensor? `GetScalar` for extension types presumably already works, not sure this is worth testing again here.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -308,19 +319,89 @@ TEST_F(TestExtensionType, TestFromTensorType) {
+
+  auto ext_arr = ExtensionType::WrapArray(ext_type, fsla_arr);
+  const auto tensor_array = std::static_pointer_cast<FixedShapeTensorArray>(ext_arr);
+  ASSERT_OK_AND_ASSIGN(const auto actual_tensor, tensor_array->ToTensor());
```
Could you perhaps add validation here? Something like:
```suggestion
  ASSERT_OK_AND_ASSIGN(const auto actual_tensor, tensor_array->ToTensor());
  ASSERT_OK(actual_tensor->Validate());
```

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -462,4 +543,121 @@ TEST_F(TestExtensionType, ToString) {
+    auto actual_ext_scalar = internal::checked_pointer_cast<ExtensionScalar>(scalar);
+    ASSERT_OK_AND_ASSIGN(auto actual_tensor,
+                         exact_ext_type->MakeTensor(actual_ext_scalar));
+    ASSERT_OK_AND_ASSIGN(auto expected_tensor,
```
Can you validate the actual tensore here and below as well?

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor_test.cc:
```diff
@@ -462,4 +543,121 @@ TEST_F(TestExtensionType, ToString) {
+  ext_type = fixed_shape_tensor(int64(), {1});
+  exact_ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(ext_type);
+  auto ext_arr = ExtensionType::WrapArray(ext_type, fsla_arr);
+  auto tensor_array = std::static_pointer_cast<FixedShapeTensorArray>(ext_arr);
```
checked_pointer_cast perhaps?

pitrou replied:
Not sure if it's an issue either, but it would be worth taking a debugger and understanding exactly what happens :-)

rok replied:
The error happened when calling `GetScalar(0)` on `FixedShapeTensorArray`. Changing the cast to:
```cpp
auto tensor_array = internal::checked_pointer_cast<ExtensionArray>(ext_arr);
```
resolved the `SIGSEGV` I was getting when doing:
```cpp
auto tensor_array = internal::checked_pointer_cast<FixedShapeTensorArray>(x);
```

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3573,17 +3602,19 @@ class FixedShapeTensorArray(ExtensionArray):
-        shape = obj.shape[1:]
-        size = obj.size / obj.shape[0]
+        shape = np.take(obj.shape, permutation)
+        values = np.ravel(obj, order="K")
```
Note that `ravel` can make a copy of the input if it is not contiguous, is this something that we want?

pitrou replied:
Hmm, I wouldn't be so sure, example:
```python
>>> obj = np.arange(24).reshape(2,3,4)[:,::2]
>>> obj
array([[[ 0,  1,  2,  3],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [20, 21, 22, 23]]])
>>> obj.strides
(96, 64, 8)
>>> permutation = (-np.array(obj.strides)).argsort(kind='stable')
>>> permutation
array([0, 1, 2])   # permutation is ok
>>> values = np.ravel(obj, order="K")
>>> values
array([ 0,  1,  2,  3,  8,  9, 10, 11, 12, 13, 14, 15, 20, 21, 22, 23])
>>> values[0] = 999
>>> values
array([999,   1,   2,   3,   8,   9,  10,  11,  12,  13,  14,  15,  20,
        21,  22,  23])
>>> obj
array([[[ 0,  1,  2,  3],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [20, 21, 22, 23]]])

# values is a copy! which we can also check using:
>>> values.ctypes.data == obj.ctypes.data
False   # base addresses are different
```

Of course, we may not really care about this, since the conversion is probably correct anyway. But we may want to add a mention in the docstring that the conversion is only zero-copy if the input array is contiguous.

rok replied:
Added a note to docstring: `If input array data is not contiguous a copy will be made.`.
Alternatively we could do `as_strided` and (hopefully) avoid copying:
```suggestion
        values = np.lib.stride_tricks.as_strided(
            obj, shape=(obj.size,), strides=(obj.dtype.itemsize,))
```

pitrou replied:
`as_strided` can be a later PR if desired. The docstring addition is good for now!

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3573,17 +3602,19 @@ class FixedShapeTensorArray(ExtensionArray):
-            fixed_shape_tensor(arrow_type, shape),
-            FixedSizeListArray.from_arrays(np.ravel(obj, order='C'), size)
+            fixed_shape_tensor(arrow_type, shape[1:], permutation=permutation[1:] - 1),
+            FixedSizeListArray.from_arrays(values, obj[0].size)
```
Note that `obj[0].size` will fail if `obj` is a 0-element array. Instead, you could use `shape[1:].prod()`.

pitrou commented on python/pyarrow/tests/test_extension_type.py:
```diff
@@ -1318,39 +1318,101 @@ def test_tensor_type():
+    arr = np.array([[1, 2, 3, 4, 5, 6], [7, 8, 9, 10, 11, 12]], dtype=value_type)
+    expected = arr[1:]
+    result = pa.FixedShapeTensorArray.from_numpy_ndarray(arr)[1:].to_numpy_ndarray()
+    np.testing.assert_array_equal(result, expected)
```
Can we add a few interesting cases?
* 1-dimensional ndarray
* 0-dimensional ndarray (should raise an error?)
* n-dimensional ndarray with shape (0, x, y)
* n-dimensional ndarray with shape (x, 0, y)

pitrou replied:
Thank you!

pitrou reviewed (changes requested):
Sorry for the delay @rok . This is starting to look good, some suggestions for improvement below :-)

rok reviewed (commented):
Thanks a lot for the review @pitrou ! Unfortunately I couldn't work on it immediately.
I think I addressed all of your points.

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -18,7 +18,9 @@
 #include <numeric>
 #include <sstream>
 
+#include <arrow/scalar.h>
```
Can we avoid the gratuitous style inconsistency?
```suggestion
#include "arrow/scalar.h"
```

(will need to reformat afterthis)

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -86,7 +88,7 @@ bool FixedShapeTensorType::ExtensionEquals(const ExtensionType& other) const {
     return false;
   }
-  const auto& other_ext = static_cast<const FixedShapeTensorType&>(other);
+  const auto& other_ext = dynamic_cast<const FixedShapeTensorType&>(other);
```
Did it fail otherwise? I'm a bit curious to know if the change to `dynamic_cast` was necessary.

(ideally this should be a `checked_cast`: i.e. dynamically checked in debug mode, not in release mode)

rok replied:
I forget what the reason was, I vaguely  remember it had to do with a change I later reverted.
Changed to `checked_cast` and it seems to work fine!

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -199,10 +201,52 @@ std::shared_ptr<Array> FixedShapeTensorType::MakeArray(
   DCHECK_EQ(data->type->id(), Type::EXTENSION);
   DCHECK_EQ("arrow.fixed_shape_tensor",
-            static_cast<const ExtensionType&>(*data->type).extension_name());
+            dynamic_cast<const ExtensionType&>(*data->type).extension_name());
```
Same question, I suppose the answer will be similar :-)

pitrou commented on cpp/src/arrow/extension/fixed_shape_tensor.cc:
```diff
@@ -293,53 +337,71 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
+  // array and we assume the first dimension will always have the greatest stride, so it
+  // will get permutation index 0 and remaining values from ext_type->permutation() need
+  // to be shifted to fill the [1, ndim+1) range. Computed permutation will be used to
+  // generate the new tensor's shape, strides and dim_names.
```
+1, thank you!

pitrou commented on cpp/src/arrow/extension/tensor_internal.h:
```diff
@@ -0,0 +1,41 @@
+
+#pragma once
+
+#include "arrow/util/print.h"
```
Should also include the necessary standard and Arrow headers, e.g.:
```suggestion
#include <cstdint>
#include <vector>

#include "arrow/status.h"
#include "arrow/util/print.h"
```

pitrou reviewed (commented):
Thanks a lot for the update. This is really getting good, just a couple comments on specific points.

rok commented:
> Thanks a lot for the update. This is really getting good, just a couple comments on specific points.

Thanks for the helpful review @pitrou, I'm happy to see this moving forward! I've addressed your points, please let me know if more changes are needed.

pitrou commented on python/pyarrow/array.pxi:
```diff
@@ -3570,27 +3570,57 @@ class FixedShapeTensorArray(ExtensionArray):
+        shape tensor.
+
+        The conversion is zero-copy.
+
```
Nit: remove extraneous empty line.
```suggestion
```

pitrou commented:
@github-actions crossbow submit -g python -g wheel

rok commented:
Added an issue for the docs: https://github.com/apache/arrow/issues/39998
Thanks for your reviews @pitrou @jorisvandenbossche @AlenkaF @alippai !

jorisvandenbossche commented on python/pyarrow/array.pxi:
```diff
@@ -3570,11 +3570,14 @@ cdef class FixedShapeTensorArray(ExtensionArray):
         and the rest of the dimensions will match the permuted shape of the fixed
         shape tensor.
 
+        The conversion is zero-copy.
```
Small nit: this is only if the conversion to numpy is zero-copy (i.e. primitive numeric data without nulls)

rok replied:
Good point, added to `VariableShapeTensor` PR https://github.com/apache/arrow/pull/38008/commits/8ca3bf7b36daf11c63a6dd665a9547997dfe3687
